rinex-cli/src/positioning/interp/time.rs

```diff
@@ -0,0 +1,220 @@
+use crate::cli::Context;
+use gnss_rtk::prelude::{Epoch, SV};
+use std::collections::HashMap;
+
+use super::Buffer as BufferTrait;
+
+struct Buffer {
+    inner: Vec<(Epoch, f64)>, //TODO
+}
+
+impl BufferTrait<f64> for Buffer {
+    fn malloc(size: usize) -> Self {
+        Self {
+            inner: Vec::with_capacity(size),
+        }
+    }
+    fn len(&self) -> usize {
+        self.inner.len()
+    }
+    /// Fill internal Buffer, which does not tolerate data gaps
+    fn at_capacity(&self) -> bool {
+        self.inner.len() == 3
+    }
+    fn clear(&mut self) {
+        self.inner.clear();
+    }
+    fn get(&self, idx: usize) -> Option<&(Epoch, f64)> {
+        self.inner.get(idx)
+    }
+    fn latest(&self) -> Option<&(Epoch, f64)> {
+        self.inner.last()
+    }
+    fn push(&mut self, x_j: (Epoch, f64)) {
+        if self.len() == 3 {
+            let mut index = 0;
+            self.inner.retain(|r| {
+                index += 1;
+                index > 1
+            });
+        }
+        self.inner.push(x_j);
+    }
+    fn snapshot(&self) -> &[(Epoch, f64)] {
+        &self.inner
+    }
+}
+
+/// Orbital state interpolator
+pub struct Interpolator<'a> {
+    ptr: Option<Epoch>,
+    buffers: HashMap<SV, Buffer>,
+    iter: Box<dyn Iterator<Item = (Epoch, SV, f64)> + 'a>,
+}
+
+impl<'a> Interpolator<'a> {
+    pub fn from_ctx(ctx: &'a Context) -> Self {
+        Self {
+            ptr: None,
+            buffers: HashMap::with_capacity(4),
+            iter: if let Some(clk) = ctx.data.clock() {
+                Box::new(
+                    clk.precise_sv_clock()
+                        .map(|(t, sv, _, prof)| (t, sv, prof.bias)),
+                )
+            } else if let Some(sp3) = ctx.data.sp3() {
+                Box::new(sp3.sv_clock())
+            } else {
+                panic!("sp3 or clock rinex currently required");
+            },
+        }
+    }
+    fn push_data(&mut self, t: Epoch, sv: SV, dt: f64) {
+        if let Some(buf) = self.buffers.get_mut(&sv) {
+            buf.fill((t, dt));
+        } else {
+            let mut buf = Buffer::malloc(3);
+            buf.fill((t, dt));
+            self.buffers.insert(sv, buf);
+        }
+        self.ptr = Some(t);
+    }
+    // Consumses n epoch
+    fn consume(&mut self, total: usize) {
```

Comment:
I can't easily follow this, is it a push-based https://doc.rust-lang.org/std/vec/struct.Vec.html#method.chunks_exact?

Reply:
It is pull based (if I follow you correctly), meaning that I pull one epoch at a time from the iterator

Reply:
I don't see how we could use methods like chunk_exact because we have no idea how many times we need to pull to complete at least one epoch. One Epoch is made of several SV and this number varies with time, as SV go out of sight